### DISCUSSION TITLE
HTTP Boot and static IP (snponly.efi)

### DISCUSSION THREAD

**Discussion by champtar** [ipxe/ipxe#938] [Q&A] [answered]
I'm playing with a Dell server (R640 / BIOS 2.17.1 / UEFI Compliance Version 2.7)
I can HTTP Boot iPXE snponly.efi (just built 1d1cf74a5e58811822bee4b3da3cff7282fcdfca) using a static IP / URL configured in the BIOS,
I see only 1 NIC as expected (net0 / type NII, MAC address matches), but even if I `ifopen net0`,
iPXE doesn't have the IP/Mask/Gateway/URL used to download iPXE.

Is snponly.efi supposed to get static IPs informations ? (haven't tried with DHCP)

**Comment by NiKiZe** [comment]
iPXE does support cached DHCP information, but probably not static IP information on that interface.

My guess is that this requires a different method for finding the up settings, but @mcb30 will have to confirm if this is the case.

**Reply by mcb30** [reply, marked as answer]
This is correct: there is no support in iPXE for pulling IP configuration from anything other than a cached DHCP packet.

The UEFI firmware **might** expose the IP configuration via the loaded device path visible to iPXE, but this would require new code in iPXE to extract the settings.

**Reply by champtar** [reply]
Thanks @NiKiZe @mcb30 
@mcb30 I see the emphasize on 'might', you are not sure what the spec says, or you are not sure the implementers follow the spec ?
any iPXE debug I can enable to check if the information is exposed ?
(if not I can forget about this method and go back to generate dynamically ipxe binaries)

**Reply by stappersg** [reply]
**might** as in _there is change that IP configuration is exposed_.

But do known that the actual message is _this would require new code in iPXE to extract the settings_.  I do read that message as _that would require new code in iPXE to extract those settings_.

> any iPXE debug I can enable to check if the information is exposed ?

...


> if not I can forget about this method and go back to generate dynamically ipxe binaries

Acknowledge

**Reply by mcb30** [reply]
> Thanks @NiKiZe @mcb30 @mcb30 I see the emphasize on 'might', you are not sure what the spec says, or you are not sure the implementers follow the spec ? any iPXE debug I can enable to check if the information is exposed ? (if not I can forget about this method and go back to generate dynamically ipxe binaries)

Using `DEBUG=efi_init` will show a message starting `EFI image device path ...` - you should be able to see if all of the relevant IP configuration information is present in this device path.

**Reply by champtar** [reply]
On Dell I get something like `PciRoot(0x1)/Pci(0x2,0x0)/Pci(0x0,0x0)/MAC(REDACTED,0x1)/IPv4(0.0.0.0)/Uri(http://REDACTED/snponly.efi)`, so this is not going to work.
Thanks for the answers and for iPXE

**Reply by NiKiZe** [reply]
I think the QEMU efi firmware has support for static IP as well, which should make it "easy" to develop a solution for that, if at all possible.

**Reply by champtar** [reply]
My use case is to deploy OS on servers in network that I don't control, that might or might not have DHCP, might or might not have internet access. If possible I want to have static files only on an http(s) server, and orchestrate everything using Ansible. I have a working solution for Dell and HPE that involves booting a custom ISO via the virtual CD functionality, and looking using Dell / HPE tools at the uri of the ISO at the query params in which I put all the networks info, so at the end I have a Live Linux system booted with a static IP, and using Ansible I can dd images or kexec into kickstart installs.

We could actually do something similar here, look at the query params for some iPXE script encoded in base64 (`.../snponly.efi?script=BASE64`), but that's just a hack that will fail with some vendor.
